pkg/querier/tail.go

```diff
@@ -1,92 +1,285 @@
+	for addr := range t.querierTailClients {
+		connectedIngestersAddr = append(connectedIngestersAddr, addr)
 	}
 
-	return true
+	newConnections, err := t.tailDisconnectedIngesters(connectedIngestersAddr)
+	if err != nil {
+		return err
+	}
+
+	if len(newConnections) != 0 {
+		for addr, tailClient := range newConnections {
+			t.querierTailClients[addr] = tailClient
+			go t.readTailClient(addr, tailClient)
+		}
+	}
+	return nil
 }
 
-func (t *tailIterator) Entry() logproto.Entry {
-	entry := t.entryIterator.Entry()
-	t.queryRequest.Start = entry.Timestamp.Add(1 * time.Nanosecond)
-	return entry
+// removes disconnected tail client from map
+func (t *Tailer) dropTailClient(addr string) {
+	t.querierTailClientsMtx.Lock()
+	defer t.querierTailClientsMtx.Unlock()
+
+	delete(t.querierTailClients, addr)
 }
 
-func (t *tailIterator) Error() error {
-	return t.err
+// keeps reading streams from grpc connection with ingesters
+func (t *Tailer) readTailClient(addr string, querierTailClient logproto.Querier_TailClient) {
+	var resp *logproto.TailResponse
+	var err error
+	defer t.dropTailClient(addr)
+	for {
+		if t.stopped {
+			if err := querierTailClient.CloseSend(); err != nil {
+				level.Error(util.Logger).Log("Error closing gprc tail client", fmt.Sprintf("%v", err))
+			}
+			break
+		}
+		resp, err = querierTailClient.Recv()
+		if err != nil {
+			level.Error(util.Logger).Log("Error receiving response from gprc tail client", fmt.Sprintf("%v", err))
+			break
+		}
+		t.pushTailResponseFromIngester(resp)
+	}
 }
 
-func (t *tailIterator) Labels() string {
-	return t.entryIterator.Labels()
+// pushes new streams from ingesters synchronously
+func (t *Tailer) pushTailResponseFromIngester(resp *logproto.TailResponse) {
+	t.streamMtx.Lock()
+	defer t.streamMtx.Unlock()
+
+	t.openStreamIterator.Push(iter.NewStreamIterator(resp.Stream))
+	if resp.DroppedStreams != nil {
+		for idx := range resp.DroppedStreams {
+			heap.Push(t.droppedStreamsIterator, *resp.DroppedStreams[idx])
+		}
+	}
 }
 
-func (t *tailIterator) Close() error {
-	return t.entryIterator.Close()
+// finds oldest entry by peeking at open stream iterator and dropped stream iterator.
+// if open stream iterator has oldest entry then pop it for sending it to tail client
+// else pop dropped stream details, to query from ingester and store.
+// Response from ingester and store is pushed to open stream for further processing
+func (t *Tailer) next() bool {
+	t.streamMtx.Lock()
+	defer t.streamMtx.Unlock()
+
+	// if we don't have any entries or any of the entries are not older than now()-delay then return false
+	if !((t.openStreamIterator.Len() != 0 && time.Now().After(t.openStreamIterator.Peek().Add(t.delayFor))) || (t.droppedStreamsIterator.Len() != 0 && time.Now().After(t.droppedStreamsIterator.Peek().Add(t.delayFor)))) {
+		return false
+	}
+
+	// If any of the dropped streams are older than open streams, pop dropped stream details for querying them
```

Comment:
I would say simplify things by removing this logic for handling dropped streams. If we need to go back and query for dropped streams, we can add that later. Let's get this working first.